keps/sig-node/1948-deallocate-dp-api/README.md

```diff
@@ -0,0 +1,840 @@
+publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to
+mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes adding two extra API calls:
+
+- `Deallocate`: (Optional). This API call is the opposite of `Allocate` and is
+needed to inform device plugins that some devices are no longer being used.
+- `PostStopContainer`: (Optional). This API call allows the device plugins to do
+ device cleanup, driver unloading, and any other actions that may be needed.
+
+Since both additions are optional, existing device plugins should continue
+functioning correctly with no needed modifications. Only device plugins that
+wish to utilize the new API calls will need to be modified.
+
+## Motivation
+
+The following are some use cases and motivations for the proposed change:
+
+- `PostStopContainer`:
+
+  - For use with some devices like FPGAs. Devices like these will need to be
+  cleaned up (i.e. de-programmed) after each use. Otherwise, they run the
+  possibility of 2 risks:
+    - If whatever is programmed on the FPGA is not cleaned up, it will keep
+    running and consuming power for no reason, on a large scale (datacenter
+    scale) this is unacceptable.
+    - If whatever is programmed on the FPGA has network access, it runs the risk
+    of continuing to send and respond to packets and pollute the network.
+  - For dynamically binding/unbinding drivers for the devices as needed.
+
+- `Deallocate`:
+
+  - For use with complex device plugins that require tracking the state of their
+  devices and learning when they are no longer in use. Multi modal devices can
+  operate in more than one mode of operation, and thus have to be advertised by
+  the device plugin as two separate devices, and the device plugin has to take
+  care to stop advertising a device when its being used in the other mode. An
+  example of multi modal devices is also using FPGAs in the following 2 modes:
+    - Use the entire FPGA as a device
+    - Split the FPGA between multiple users, essentially advertising one FPGA as
+    multiple smaller FPGAs.
+    - A device plugin needs to know when a full FPGA will stop being used so it
+    can go back to advertise the FPGA partitions, and vice versa.
+  - To maintain the same logical splitting of `Allocate` and `PreStartContainer`
+
+### Goals
+
+- Add the `PostStopContainer` and `Deallocate` API calls to the device plugin
+API.
+- Make the new added API calls optional, as they are not needed for all devices.
+- Maintain compatibility with existing device plugins.
+
+### Non-Goals
+
+- Make any modifications to the main API calls of the device plugin API.
+- Make changes specific to one type of devices.
+
+## Use Cases
```

Comment:
Another use case is "shared devices", The device plugin assumes the allocated devices are local to the node but that doesn't necessarily have to be the case. If the discovered devices are shared across nodes, then a `Deallocate(..)` is necessary to guarantee the resources are quickly freed for others to use (This explains [Akri](https://github.com/project-akri/akri) usecase)

Reply:
Thanks @romoh added to the list of use-cases.